MSAL/src/native_auth/public/state_machine/delegate/SignUpDelegateDispatchers.swift

```diff
@@ -0,0 +1,201 @@
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import Foundation
+
+final class SignUpPasswordStartDelegateDispatcher: DelegateDispatcher<SignUpPasswordStartDelegate> {
+
+    func dispatchSignUpCodeRequired(
+        newState: SignUpCodeRequiredState,
+        sentTo: String,
+        channelTargetType: MSALNativeAuthChannelType,
+        codeLength: Int
+    ) async {
+        if let signUpCodeRequired = delegate.onSignUpCodeRequired {
+            telemetryUpdate?(.success(()))
+            await signUpCodeRequired(newState, sentTo, channelTargetType, codeLength)
+        } else {
+            let error = SignUpPasswordStartError(type: .generalError, message: errorMessage(for: "onSignUpCodeRequired"))
+            telemetryUpdate?(.failure(error))
+            await delegate.onSignUpPasswordError(error: error)
+        }
+    }
+
+    func dispatchSignUpAttributesInvalid(attributeNames: [String]) async {
+        if let signUpAttributes = delegate.onSignUpAttributesInvalid {
+            telemetryUpdate?(.success(()))
+            await signUpAttributes(attributeNames)
+        } else {
+            let error = SignUpPasswordStartError(type: .generalError, message: errorMessage(for: "onSignUpAttributesInvalid"))
+            telemetryUpdate?(.failure(error))
+            await delegate.onSignUpPasswordError(error: error)
+        }
+    }
+}
+
+final class SignUpStartDelegateDispatcher: DelegateDispatcher<SignUpStartDelegate> {
+
+    func dispatchSignUpCodeRequired(
+        newState: SignUpCodeRequiredState,
+        sentTo: String,
+        channelTargetType: MSALNativeAuthChannelType,
+        codeLength: Int
+    ) async {
+        if let signUpCodeRequired = delegate.onSignUpCodeRequired {
```

Comment:
Curious to know if you considered `guard let` instead of `if let` for all the dispatcher methods?  I think it would be clearer/more swifty.

Reply:
I considered it, but if in the future we have more delegates, it will have to be changed to an `if/else if/else`. 
```swift
if let onSignUpCompletedNew = delegate.onSignUpCompletedNew {
  await onSignUpCompletedNew(newState, date)
} else if let onSignUpCompletedOld = delegate.onSignUpCompletedOld {
  await onSignUpCompletedOld(newState)
} else {
  await delegate.onSignUpError(error: error)
}
```

By using if/let now, we have the whole class written in the same style, even if we have newer delegates in the future.

Reply:
Thanks that makes a lot of sense. I knew you would have a good reason 😀